algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c

```diff
@@ -0,0 +1,92 @@
+
+//Circular Linked List
+//insertion and Deletion at the beginning of the Linked List
+#include<stdio.h>
+#include<stdlib.h>
+
+//  Structure Definition
+
+// 1. Data of Integer type.
+// 2. Pointer to the Next Node .
+struct node {
+    int data ;
+    struct node *next ;
+} ;
+struct node *head = NULL ;
+//Use tail pointer to avoid traversing the Linked List while inserting of the new created node
+//because after every insertion of a new node we need to point make the last node point to the newly created Node 
+struct node *tail = NULL ;
+
+//Function to Print the Linked List
+void printList(){
+    printf("\nItem in the Linked List are :  ");
+    // Initialize the ptr(pointer) with head location .
+    struct node *ptr = head ;
+    do{
+       printf("%d ",ptr->data) ;  //Print the value of the Node
+       ptr = ptr->next ;         //Increment the ptr 
+    }while(ptr!=head);           //// While Loop until we encounter a Node pointer ptr not pointing to the same location as head .
+}
+
+//Insertion at the Beginning of the Circular Linked List
+
+void insertBeg(int data){
+ //Creating a link
+ //Malloc is for space allocation .
+ struct node *new_node = (struct node*) malloc(sizeof(struct node)) ;
+ if(new_node==NULL){
+     printf("Cannot create the Node");
+ }
+ else{
+     
+     new_node->data=data;
+     if(head==NULL){
+     	//making the head and the tail pointer to point to the newly created node
+         head = new_node ;
+         tail = new_node;
+     }
+     else{
+        //Point the new_node pointer to the current head
+        new_node->next = head ;
+        head = new_node ;
+     }
+     //making the tail pointer to point to the new head
+    tail->next=head;
+     
+     
+ }
+ 
+}
+
+//Deletion at the Beginning of the Circular Linked List
+
+void deleteBeg(){
+  //NO Deletion when list is empty
+   if(head == NULL) {
+    printf("List is Empty , deletion not possible") ;
+   }
+  //Store head location in a temporary Node pointer.
+  struct node *temp = head ;
```

Comment:
What will happen when there is only one element left in the list ?

Reply:
if there is only one element left in the list then both the head and the tail pointer will point to the that same Node
then we create a temporary pointer [temp] and and make it point to the head
we delete the node pointed by temp
next we make the head point to head->next which is null
Now, head will point to NULL and temp will be deleted
I don't see what is the problem here sir??

Reply:
>we delete the node pointed by temp
next we make the head point to head->next which is null

here is the problem,
`free()` just deallocates the memory and doesn't make the pointer point to `NULL`, the pointer will still be pointing to that particular memory location only, it's just that that now that memory location would not be accessible to your program anymore and an attempt to do so ( which will happen when you try to print the contents via `printList()` function ) will lead to [undefined behaviour](https://en.wikipedia.org/wiki/Undefined_behavior) ( you can see this happening when you attempt to perform more deletions than the size of the list ) , such pointers are known as [dangling pointers](https://en.wikipedia.org/wiki/Dangling_pointer) and best way to avoid it is to handle the case when there is only one element in the list separately and make the head manually point to `NULL` after deleting it.

you can do something like this 
```c
if(head == tail)
  {
      free(head);
      head = NULL;
      return;
  }
```

---

Also you might have seen till now that your `printList()` function is also unsafe as it will not check if the list is empty or not and will still try to access the list even if it's empty ( head is pointing to `NULL` ), you might wanna change that.

soemthing like this 
```c
if(head == NULL)
  {
      printf("\nList is empty");
      return;
  }
```

Reply:
You can actually test it for yourself, try inserting only one element to the list and call `deleteBeg()` and then `printList()`
you will a get a `segmentation fault` error or an unexpected behaviour

Reply:
@Ashborn-SM @Arsenic-ATG  Thank you sir for pointing my mistake, I feel like solving a problem on Hackerrank where i pass the test case but when i submit the code i fall short because i did not handle the corner case haha
I have also updated the README.md file as you suggested
if there is any other mistake please do tell me
Thank you again for guiding me